File: cpp/test/TestTokenizer.cpp

#include "Tokenizer.h"
#include "ParserSession.h"
#include "TokenEnumRegistration.h"

#include "gtest/gtest.h"

class ParserSession;

using ParserSessionPtr = ParserSession *;


class TokenizerTest : public ::testing::Test {
protected:
    
    static ParserSessionPtr session;
    
    static void SetUpTestSuite() {
        
        session = new ParserSession();
    }
    
    static void TearDownTestSuite() {
        
        delete session;
    }
    
    void SetUp() override {
        
    }
    
    void TearDown() override {
        
        session->deinit();
    }
    
};

ParserSessionPtr TokenizerTest::session;


//
// This was asserting
//
TEST_F(TokenizerTest, Bug1) {
    
    auto strIn = std::string("\\.GG");
    
    auto str = reinterpret_cast<Buffer>(strIn.c_str());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(str, strIn.size(), nullptr, opts);
    
    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    SUCCEED();
}

//
// This used to assert
//
TEST_F(TokenizerTest, Bug2) {
    
    auto strIn = std::string("<<<");
    
    auto str = reinterpret_cast<Buffer>(strIn.c_str());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(str, strIn.size(), nullptr, opts);
    
    auto Tok = Tokenizer_currentToken(session, TOPLEVEL);
    
    Tok.skip(session);
    
    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    SUCCEED();
}

//
// This used to assert
//
TEST_F(TokenizerTest, Bug3) {
    
    auto strIn = std::string("\\\r");
    
    auto str = reinterpret_cast<Buffer>(strIn.c_str());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(str, strIn.size(), nullptr, opts);
    
    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    SUCCEED();
}

TEST_F(TokenizerTest, Bug4) {
    
    auto strIn = std::string("\\[");
    
    auto str = reinterpret_cast<Buffer>(strIn.c_str());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(str, strIn.size(), nullptr, opts);
    
    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    SUCCEED();
}

TEST_F(TokenizerTest, Bug5) {
    
    auto strIn = std::string("\"a\\\\\r\nb\"");

    auto str = reinterpret_cast<Buffer>(strIn.c_str());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(str, strIn.size(), nullptr, opts);

    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    SUCCEED();
}

TEST_F(TokenizerTest, IntegerRealMixup) {
    
    auto strIn = std::string("0..");

    auto str = reinterpret_cast<Buffer>(strIn.c_str());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(str, strIn.size(), nullptr, opts);

    auto Tok1 = Tokenizer_currentToken(session, TOPLEVEL);

    EXPECT_EQ(Tok1, Token(TOKEN_INTEGER, session->start + 0, 1, Source(SourceLocation(1, 1), SourceLocation(1, 2))));

    Tok1.skip(session);

    auto Tok2 = Tokenizer_currentToken(session, TOPLEVEL);

    EXPECT_EQ(Tok2, Token(TOKEN_DOTDOT, session->start + 1, 2, Source(SourceLocation(1, 2), SourceLocation(1, 4))));

    Tok2.skip(session);

    auto Tok3 = Tokenizer_currentToken(session, TOPLEVEL);

    EXPECT_EQ(Tok3, Token(TOKEN_ENDOFFILE, session->start + 3, 1, Source(SourceLocation(1, 4), SourceLocation(1, 4))));
    
    EXPECT_EQ(session->nonFatalIssues.size(), 1u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
}

TEST_F(TokenizerTest, Basic2) {
    
    auto strIn = std::string("\\[Alpha]bc+1");

    auto str = reinterpret_cast<Buffer>(strIn.c_str());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(str, strIn.size(), nullptr, opts);

    auto Tok1 = Tokenizer_currentToken(session, TOPLEVEL);

    EXPECT_EQ(Tok1, Token(TOKEN_SYMBOL, session->start + 0, 10, Source(SourceLocation(1, 1), SourceLocation(1, 11))));

    Tok1.skip(session);

    auto Tok2 = Tokenizer_currentToken(session, TOPLEVEL);

    EXPECT_EQ(Tok2, Token(TOKEN_PLUS, session->start + 10, 1, Source(SourceLocation(1, 11), SourceLocation(1, 12))));

    Tok2.skip(session);

    auto Tok3 = Tokenizer_currentToken(session, TOPLEVEL);

    EXPECT_EQ(Tok3, Token(TOKEN_INTEGER, session->start + 11, 1, Source(SourceLocation(1, 12), SourceLocation(1, 13))));

    Tok3.skip(session);

    auto Tok4 = Tokenizer_currentToken(session, TOPLEVEL);

    EXPECT_EQ(Tok4, Token(TOKEN_ENDOFFILE, session->start + 12, 1, Source(SourceLocation(1, 13), SourceLocation(1, 13))));
    
    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
}

TEST_F(TokenizerTest, OldAssert1) {
    
    auto strIn = std::string("8*");

    auto str = reinterpret_cast<Buffer>(strIn.c_str());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(str, strIn.size(), nullptr, opts);

    auto Tok = Tokenizer_currentToken(session, TOPLEVEL);

    EXPECT_EQ(Tok, Token(TOKEN_INTEGER, session->start, 1, Source(SourceLocation(1, 1), SourceLocation(1, 2))));
    
    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
}

TEST_F(TokenizerTest, Basic3) {
    
    auto strIn = std::string("{\n}");

    auto str = reinterpret_cast<Buffer>(strIn.c_str());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(str, strIn.size(), nullptr, opts);

    auto Tok = Tokenizer_currentToken(session, TOPLEVEL);

    EXPECT_EQ(Tok, Token(TOKEN_OPENCURLY, session->start, 1, Source(SourceLocation(1, 1), SourceLocation(1, 2))));

    Tok.skip(session);

    //
    // Clear 0x100 because we are inside a group now
    //
    Tok = Tokenizer_currentToken(session, TOPLEVEL & ~(RETURN_TOPLEVELNEWLINE));

    EXPECT_EQ(Tok, Token(TOKEN_INTERNALNEWLINE, session->start + 1, 1, Source(SourceLocation(1, 2), SourceLocation(2, 1))));

    Tok.skip(session);

    Tok = Tokenizer_currentToken(session, TOPLEVEL);

    EXPECT_EQ(Tok, Token(TOKEN_CLOSECURLY, session->start + 2, 1, Source(SourceLocation(2, 1), SourceLocation(2, 2))));

    Tok.skip(session);
    
    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
}

TEST_F(TokenizerTest, Basic4) {
    
    const unsigned char arr[] = { 0xff };
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(arr, 1, nullptr, opts);
    
    EXPECT_EQ(session->SrcLoc, SourceLocation(1, 1));
    
    auto Tok = Tokenizer_currentToken(session, TOPLEVEL);
    
    EXPECT_EQ(Tok, Token(TOKEN_ERROR_UNSAFECHARACTERENCODING, session->start + 0, 1, Source(SourceLocation(1, 1), SourceLocation(1, 2))));
    
    EXPECT_EQ(session->SrcLoc, SourceLocation(1, 1));
    
    Tok.skip(session);
    
    Tok = Tokenizer_currentToken(session, TOPLEVEL);
    
    EXPECT_EQ(Tok, Token(TOKEN_ENDOFFILE, session->start + 1, 1, Source(SourceLocation(1, 2), SourceLocation(1, 2))));
    
    Tok.skip(session);
    
    EXPECT_EQ(session->SrcLoc, SourceLocation(1, 2));
    
    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 1u);
}

TEST_F(TokenizerTest, Crash1) {
    
    const unsigned char arr[] = { '6', '`', '5', '.', '.' };
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(arr, 5, nullptr, opts);
    
    Tokenizer_currentToken(session, TOPLEVEL);
    
    EXPECT_EQ(session->nonFatalIssues.size(), 1u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    SUCCEED();
}

TEST_F(TokenizerTest, LineContinuation1) {

    auto strIn = std::string("ab\\\ncd");

    auto str = reinterpret_cast<Buffer>(strIn.c_str());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(str, strIn.size(), nullptr, opts);

    auto Tok = Tokenizer_currentToken(session, TOPLEVEL);

    EXPECT_EQ(Tok, Token(TOKEN_SYMBOL, session->start + 0, 6, Source(SourceLocation(1, 1), SourceLocation(2, 3))));
    
    Tokenizer_nextToken(session, TOPLEVEL);
    
    Tok = Tokenizer_currentToken(session, TOPLEVEL);

    EXPECT_EQ(Tok, Token(TOKEN_ENDOFFILE, session->start + 6, 1, Source(SourceLocation(2, 3), SourceLocation(2, 3))));
    
    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
}

TEST_F(TokenizerTest, LineContinuation2) {
    
    auto strIn = std::string("ab\\\r\ncd");

    auto str = reinterpret_cast<Buffer>(strIn.c_str());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(str, strIn.size(), nullptr, opts);

    auto Tok = Tokenizer_currentToken(session, TOPLEVEL);

    EXPECT_EQ(Tok, Token(TOKEN_SYMBOL, session->start + 0, 7, Source(SourceLocation(1, 1), SourceLocation(2, 3))));
    
    Tokenizer_nextToken(session, TOPLEVEL);
    
    Tok = Tokenizer_currentToken(session, TOPLEVEL);

    EXPECT_EQ(Tok, Token(TOKEN_ENDOFFILE, session->start + 7, 1, Source(SourceLocation(2, 3), SourceLocation(2, 3))));
    
    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
}

TEST_F(TokenizerTest, LineContinuation3) {
    
    auto strIn = std::string("ab\\\rcd");
    
    auto str = reinterpret_cast<Buffer>(strIn.c_str());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(str, strIn.size(), nullptr, opts);

    auto Tok = Tokenizer_currentToken(session, TOPLEVEL);

    EXPECT_EQ(Tok, Token(TOKEN_SYMBOL, session->start + 0, 6, Source(SourceLocation(1, 1), SourceLocation(2, 3))));

    Tokenizer_nextToken(session, TOPLEVEL);
    
    Tok = Tokenizer_currentToken(session, TOPLEVEL);

    EXPECT_EQ(Tok, Token(TOKEN_ENDOFFILE, session->start + 6, 1, Source(SourceLocation(2, 3), SourceLocation(2, 3))));
    
    EXPECT_EQ(session->nonFatalIssues.size(), 1u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
}

TEST_F(TokenizerTest, LineContinuation4) {
    
    auto strIn = std::string("1\\\n");
    
    auto str = reinterpret_cast<Buffer>(strIn.c_str());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(str, strIn.size(), nullptr, opts);

    auto Tok = Tokenizer_currentToken(session, TOPLEVEL);

    EXPECT_EQ(Tok, Token(TOKEN_INTEGER, session->start + 0, 1, Source(SourceLocation(1, 1), SourceLocation(1, 2))));
    
    Tokenizer_nextToken(session, TOPLEVEL);
    
    Tok = Tokenizer_currentToken(session, TOPLEVEL);

    EXPECT_EQ(Tok, Token(TOKEN_ENDOFFILE, session->start + 1, 3, Source(SourceLocation(1, 2), SourceLocation(2, 1))));
    
    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
}

File: crate/src/tests/test_tokenizer.rs
use crate::{
    source::{ByteSpan, NextPolicyBits::RETURN_TOPLEVELNEWLINE, SourceLocation, TOPLEVEL},
    token::Token,
    token_enum::TokenEnum::*,
    tokenizer::{Tokenizer_currentToken, Tokenizer_nextToken},
    EncodingMode, FirstLineBehavior, ParserSession, SourceConvention, DEFAULT_TAB_WIDTH,
};


//
// This was asserting
//
#[test]
fn TokenizerTest_Bug1() {
    let strIn = "\\.GG";

    let session = ParserSession::new(
        strIn.as_bytes(),
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);
}

//
// This used to assert
//
#[test]
fn TokenizerTest_Bug2() {
    let strIn = "<<<";

    let mut session = ParserSession::new(
        strIn.as_bytes(),
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    let Tok = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    Tok.skip(&mut session.tokenizer);

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);
}

//
// This used to assert
//
#[test]
fn TokenizerTest_Bug3() {
    let strIn = "\\\r";

    let session = ParserSession::new(
        strIn.as_bytes(),
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);
}

#[test]
fn TokenizerTest_Bug4() {
    let strIn = "\\[";

    let session = ParserSession::new(
        strIn.as_bytes(),
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);
}

#[test]
fn TokenizerTest_Bug5() {
    let strIn = "\"a\\\\\r\nb\"";

    let session = ParserSession::new(
        strIn.as_bytes(),
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);
}

#[test]
fn TokenizerTest_IntegerRealMixup() {
    let strIn = "0..";

    let mut session = ParserSession::new(
        strIn.as_bytes(),
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    let Tok1 = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    assert_eq!(
        Tok1,
        Token::new3(TOKEN_INTEGER, ByteSpan::new(0, 1), src!(1:1-1:2))
    );

    Tok1.skip(&mut session.tokenizer);

    let Tok2 = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    assert_eq!(
        Tok2,
        Token::new3(TOKEN_DOTDOT, ByteSpan::new(1, 2), src!(1:2-1:4))
    );

    Tok2.skip(&mut session.tokenizer);

    let Tok3 = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    assert_eq!(
        Tok3,
        Token::new3(TOKEN_ENDOFFILE, ByteSpan::new(3, 0), src!(1:4-1:4))
    );

    assert_eq!(session.nonFatalIssues().len(), 1);
    assert_eq!(session.fatalIssues().len(), 0);
}

#[test]
fn TokenizerTest_Basic2() {
    let strIn = "\\[Alpha]bc+1";

    let mut session = ParserSession::new(
        strIn.as_bytes(),
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    let Tok1 = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    assert_eq!(
        Tok1,
        Token::new3(TOKEN_SYMBOL, ByteSpan::new(0, 10), src!(1:1-1:11))
    );

    Tok1.skip(&mut session.tokenizer);

    let Tok2 = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    assert_eq!(
        Tok2,
        Token::new3(TOKEN_PLUS, ByteSpan::new(10, 1), src!(1:11-1:12))
    );

    Tok2.skip(&mut session.tokenizer);

    let Tok3 = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    assert_eq!(
        Tok3,
        Token::new3(TOKEN_INTEGER, ByteSpan::new(11, 1), src!(1:12-1:13))
    );

    Tok3.skip(&mut session.tokenizer);

    let Tok4 = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    assert_eq!(
        Tok4,
        Token::new3(TOKEN_ENDOFFILE, ByteSpan::new(12, 0), src!(1:13-1:13))
    );

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);
}

#[test]
fn TokenizerTest_OldAssert1() {
    let strIn = "8*";

    let mut session = ParserSession::new(
        strIn.as_bytes(),
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    let Tok = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    assert_eq!(
        Tok,
        Token::new3(TOKEN_INTEGER, ByteSpan::new(0, 1), src!(1:1-1:2))
    );

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);
}

#[test]
fn TokenizerTest_Basic3() {
    let strIn = "{\n}";

    let mut session = ParserSession::new(
        strIn.as_bytes(),
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    let mut Tok = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    assert_eq!(
        Tok,
        Token::new3(TOKEN_OPENCURLY, ByteSpan::new(0, 1), src!(1:1-1:2))
    );

    Tok.skip(&mut session.tokenizer);

    //
    // Clear 0x100 because we are inside a group now
    //
    Tok = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL & !(RETURN_TOPLEVELNEWLINE));

    assert_eq!(
        Tok,
        Token::new3(TOKEN_INTERNALNEWLINE, ByteSpan::new(1, 1), src!(1:2-2:1))
    );

    Tok.skip(&mut session.tokenizer);

    Tok = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    assert_eq!(
        Tok,
        Token::new3(TOKEN_CLOSECURLY, ByteSpan::new(2, 1), src!(2:1-2:2))
    );

    Tok.skip(&mut session.tokenizer);

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);
}

#[test]
fn TokenizerTest_Basic4() {
    let arr = &[0xff];

    let mut session = ParserSession::new(
        arr,
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    assert_eq!(session.tokenizer.SrcLoc, SourceLocation::new(1, 1));

    assert_eq!(session.tokenizer.wasEOF, false);

    let mut Tok = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    assert_eq!(
        Tok,
        Token::new3(
            TOKEN_ERROR_UNSAFECHARACTERENCODING,
            ByteSpan::new(0, 1),
            src!(1:1-1:2)
        )
    );

    assert_eq!(session.tokenizer.SrcLoc, SourceLocation::new(1, 1));

    assert_eq!(session.tokenizer.wasEOF, false);

    Tok.skip(&mut session.tokenizer);

    Tok = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    assert_eq!(
        Tok,
        Token::new3(TOKEN_ENDOFFILE, ByteSpan::new(1, 0), src!(1:2-1:2))
    );

    Tok.skip(&mut session.tokenizer);

    assert_eq!(session.tokenizer.SrcLoc, SourceLocation::new(1, 2));

    assert_eq!(session.tokenizer.wasEOF, true);

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 1);
}

#[test]
fn TokenizerTest_Crash1() {
    let arr = &[b'6', b'`', b'5', b'.', b'.'];

    let mut session = ParserSession::new(
        arr,
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    assert_eq!(session.nonFatalIssues().len(), 1);
    assert_eq!(session.fatalIssues().len(), 0);
}

#[test]
fn TokenizerTest_LineContinuation1() {
    let strIn = "ab\\\ncd";

    let mut session = ParserSession::new(
        strIn.as_bytes(),
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    let mut Tok = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    assert_eq!(
        Tok,
        Token::new3(TOKEN_SYMBOL, ByteSpan::new(0, 6), src!(1:1-2:3))
    );

    Tokenizer_nextToken(&mut session.tokenizer, TOPLEVEL);

    Tok = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    assert_eq!(
        Tok,
        Token::new3(TOKEN_ENDOFFILE, ByteSpan::new(6, 0), src!(2:3-2:3))
    );

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);
}

#[test]
fn TokenizerTest_LineContinuation2() {
    let strIn = "ab\\\r\ncd";

    let mut session = ParserSession::new(
        strIn.as_bytes(),
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    let mut Tok = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    assert_eq!(
        Tok,
        Token::new3(TOKEN_SYMBOL, ByteSpan::new(0, 7), src!(1:1-2:3))
    );

    Tokenizer_nextToken(&mut session.tokenizer, TOPLEVEL);

    Tok = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    assert_eq!(
        Tok,
        Token::new3(TOKEN_ENDOFFILE, ByteSpan::new(7, 0), src!(2:3-2:3))
    );

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);
}

#[test]
fn TokenizerTest_LineContinuation3() {
    let strIn = "ab\\\rcd";

    let mut session = ParserSession::new(
        strIn.as_bytes(),
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    let mut Tok = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    assert_eq!(
        Tok,
        Token::new3(TOKEN_SYMBOL, ByteSpan::new(0, 6), src!(1:1-2:3))
    );

    Tokenizer_nextToken(&mut session.tokenizer, TOPLEVEL);

    Tok = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    assert_eq!(
        Tok,
        Token::new3(TOKEN_ENDOFFILE, ByteSpan::new(6, 0), src!(2:3-2:3))
    );

    assert_eq!(session.nonFatalIssues().len(), 1);
    assert_eq!(session.fatalIssues().len(), 0);
}

#[test]
fn TokenizerTest_LineContinuation4() {
    let strIn = "1\\\n";

    let mut session = ParserSession::new(
        strIn.as_bytes(),
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    let mut Tok = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    assert_eq!(
        Tok,
        Token::new3(TOKEN_INTEGER, ByteSpan::new(0, 1), src!(1:1-1:2))
    );

    Tokenizer_nextToken(&mut session.tokenizer, TOPLEVEL);

    Tok = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    assert_eq!(
        Tok,
        Token::new3(TOKEN_ENDOFFILE, ByteSpan::new(1, 2), src!(1:2-2:1))
    );

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);
}
